contracts/wrapper/NameWrapper.sol

```diff
@@ -511,27 +517,98 @@ contract NameWrapper is
+        expiry = _normaliseExpiry(expiry, oldExpiry, parentExpiry);
+
+        // Make sure that IS_DOT_ETH is not part of the fuses. 
         _checkFusesAreSettable(node, fuses);
-        bytes memory name = _saveLabel(parentNode, node, label);
-        expiry = _checkParentFusesAndExpiry(parentNode, node, fuses, expiry);
 
-        if (ownerOf(uint256(node)) == address(0)) {
+        // If parent controlled fuses in the node are being set, make sure CANNOT_UNWRAP
+        // is burned in the fuses of the parent node. 
+        if (fuses & PARENT_CONTROLLED_FUSES != 0 && parentFuses & CANNOT_UNWRAP == 0) {
+            revert OperationProhibited(node);
+        }
+
+        // Make sure the caller is the parent owner or approved, and the parent name is not expired. 
+        if ((baseOwner != msg.sender && !isApprovedForAll(baseOwner, msg.sender)) || 
+            (parentFuses & IS_DOT_ETH != 0 && parentExpiry - GRACE_PERIOD < block.timestamp)
+            ) {
+            revert Unauthorised(parentNode, msg.sender);
+        }
+
+        // We are no longer using the parent owner "baseOwner" so we now
+        // can set it to the registry owner to save stack space. 
+        baseOwner = ens.owner(node);
+
+        // If the name does not exist in the registery, check to see if it can be 
+        // created. If it does exist, then check to see if it has been wrapped.   
+        if (baseOwner == address(0)) {
+
+            // Check to make sure CANNOT_CREATE_SUBDOMAIN is not set.
+            if (parentFuses & CANNOT_CREATE_SUBDOMAIN != 0) {
+                revert OperationProhibited(node);
+            }
+
+            // If a owner controlled fuse are being burned, check to make sure 
+            // PARENT_CANNOT_CONTROL and CANNOT_UNWRAP have been burned. 
+            _canFusesBeBurned(node, fuses);
+
+            // If we pass all the checks, register the subname and wrap it. 
+            bytes memory _name = _saveLabel(parentNode, node, label);
             ens.setSubnodeOwner(parentNode, labelhash, address(this));
-            _wrap(node, name, owner, fuses, expiry);
+
+            _mintWithData(node, owner, fuses, expiry, nodeOwner, nodeFuses, oldExpiry);
+            emit NameWrapped(node, _name, owner, fuses, expiry);
+
         } else {
-            _updateName(parentNode, node, label, owner, fuses, expiry);
+
+            // If the name does exist in the registry check to make sure that 
+            // PARENT_CANNOT_CONTROL has not been burned.  
+            if (nodeFuses & PARENT_CANNOT_CONTROL != 0) {
+                revert OperationProhibited(node);
+            }
+
+            //Notice: In order to limit the stack height we are temporaily using label as "name".
+            label = string(_saveLabel(parentNode, node, label));
+
+            // If the name is not wrapped, then register the name and wrap it. 
+            if (nodeOwner == address(0)) {
```

Comment:
ownerOf(uint256(node)) => nodeOwner reduces the gas a lot for the low end: 

My numbers are now: 
min         max          average
60391  ·     131996  ·      123610
53744  ·     124301  ·      116197

(-6,647)      (-7,695)       (-7,413)